src/EffectiveCSharp.Analyzers/PreferDeclarationInitializersToAssignmentStatementsAnalyzer.cs

```diff
@@ -0,0 +1,656 @@
+                || !(assignment.Left is MemberAccessExpressionSyntax
+                || isFieldAssignment))
+            {
+                continue;
+            }
+
+            IdentifierNameSyntax identifierName = isFieldAssignment ? (IdentifierNameSyntax)assignment.Left : (IdentifierNameSyntax)((MemberAccessExpressionSyntax)assignment.Left).Name;
+
+            // We assume the identifier is a field. Calls to semantic model are super expensive, so we want to avoid them as much as possible.
+            string fieldName = identifierName.Identifier.Text;
+            if (!fields.TryGetValue(fieldName, out FieldInitializationInfo fieldInfo))
+            {
+                fieldInfo = new FieldInitializationInfo(fieldName);
+                fields.Add(fieldName, fieldInfo);
+            }
+
+            if (fieldInfo.ShouldNotInitializeInDeclaration)
+            {
+                continue;
+            }
+
+            if (assignment.Right is InvocationExpressionSyntax invocationExpressionSyntax
+                && !IsStaticMethodInvocation(invocationExpressionSyntax, context))
+            {
+                // If the assignment is to a non-static method invocation, we can ignore it.
+                fieldInfo.ShouldNotInitializeInDeclaration = true;
+                continue;
+            }
+
+            if (checkConstructorParameters)
+            {
+                IOperation? operation = context.SemanticModel.GetOperation(assignment.Right, context.CancellationToken);
+
+                // ILocalReferenceOperation: We do not yet have logic to verify if it is related to a constructor parameter, so we assume it is.
+                if (operation is IParameterReferenceOperation
+                    || operation is ILocalReferenceOperation
+                    || IsConstructorParameterInUse(assignment.Right, context))
+                {
+                    fieldInfo.ShouldNotInitializeInDeclaration = true;
+                    continue;
+                }
+            }
+
+            ProcessFieldInitializerCandidate(
+                assignment,
+                (ExpressionStatementSyntax)assignment.Parent!,
+                fieldInfo);
+        }
+    }
+
+    /// <summary>
+    /// Checks whether the invocation expression is a static method invocation.
+    /// </summary>
+    /// <param name="invocationExpression">The target <see cref="InvocationExpressionSyntax"/>.</param>
+    /// <param name="context">The <see cref="SyntaxNodeAnalysisContext"/>.</param>
+    /// <returns>A bool on whether the invocation is for a static method.</returns>
+    private static bool IsStaticMethodInvocation(InvocationExpressionSyntax invocationExpression, SyntaxNodeAnalysisContext context)
+    {
+        // Get the symbol information for the method being invoked
+        IMethodSymbol? methodSymbol = (IMethodSymbol?)context.SemanticModel.GetSymbolInfo(invocationExpression, context.CancellationToken).Symbol;
```

Comment:
Note: this is expensive

Reply:
If we don't want to check if a method is static, we could just assume any invocation in a constructor with a parameter is not safe to put in a declaration.

We could also just apply that blindly to all method calls we encounter. That would save cycles but cost us in accuracy. Thoughts?